### ISSUE TITLE
Notify admins when a user's GitHub company or organization changes

### ISSUE THREAD

**Issue by joemasilotti** [joemasilotti/railsdevs.com#779]
Continuing #756 but with GitHub...

I'm exploring another way to keep businesses honest and pay the required hiring fee. One idea is to track developers on ~LinkedIn~ GitHub and see if their job status changes.

Documented in the linked issue above, the GitHub API can fetch a lot of information for free. I'm most interested in "company" and "organizations". These most correlate to a developer getting a new job.

A PR addressing this issue should:

- [ ] Send an email to admins once per day of any GitHub changes for developers

This is left intentionally vague as there are a ton of ways to do this. I imagine a rough outline would be running a daily background job to fetch the relevant information, comparing it to existing records, and sending an email with the changes.

I asked about how folks handle this [on Twitter](https://twitter.com/joemasilotti/status/1622367065389805568?s=20&t=9wysN7G7_9Iuf1zFH7OafQ) and got a variety of responses. The most viable (in my opinion) is working with [`ActiveModel::Dirty`](https://api.rubyonrails.org/classes/ActiveModel/Dirty.html) in some way. That means we can rely on the framework to do the "diff" and not calculate it manually.

I'm open to ideas and questions!

**Comment by PatrickMcSweeny** [comment]
I looked that API documentation and it appears that the endpoint mentioned in the linked issue only returns information for the user that is currently authenticated with the API. [This endpoint](https://docs.github.com/en/rest/orgs/orgs?apiVersion=2022-11-28#list-organizations-for-a-user) returns publicly available organization memberships for a user, but this would not cover private information about organizations. Another possibility would be to use [this endpoint](https://docs.github.com/en/rest/orgs/orgs?apiVersion=2022-11-28#get-an-organization] which provides a URL to list public members of that organization.

I guess the major sticking point is how big of an issue relying on publicly available information about organization memberships would be. If @joemasilotti feels that it is worthwhile to move forward on this, I would be willing to work on it.

**Comment by joemasilotti** [comment]
Great point! I think that company plus public organization info is better than nothing right now. But let's test that assumption.

Want to throw together a throw-away PR that tallies how many folks have their company and/or an organization listed on GitHub? I can run it against production and share info here.

Based on that we can decide to move forward with a real PR that persists data, notifies admins, is tested, etc.

**Comment by PatrickMcSweeny** [comment]
Sure, I can do that. Would you like me to work with actual user data? I could probably get by with a list of GitHub usernames

**Comment by PatrickMcSweeny** [comment]
Actually, nevermind. I will just come up with something that will use test data locally.

**Comment by joemasilotti** [comment]
Exactly. Then I can pull your code to my machine and run it against a local version of production data.

**Comment by joemasilotti** [comment]
Verified by #783, over 40% of developers linking to their GitHub profile have either organization or company filled in.

Let's move forward with this PR!

I also noticed that the API has an hourly rate limit of 5000. So we should be good (for a while) in not hitting that.

**Comment by PatrickMcSweeny** [comment]
I'm thinking of adding a github_company attribute/column to the User model and creating a GithubOrganization model with an association to user. I would then add a callback to the new model that would send an email to admin(s) when a new org gets added, with the exception of when the table gets populated for the first time. I would then add a rake task that would actually hit the API and could be run as a cron job. Does that sound good?

**Comment by joemasilotti** [comment]
That sounds great with one change - let's move all of the new data to its own model. Even if that means we have a list/set/jsonb column for the organizations.

**Comment by PatrickMcSweeny** [comment]
@joemasilotti  I have made some progress, but if someone is really eager to work on this I would be willing to hand it off.

**Comment by joemasilotti** [comment]
OK, cool. I'm going to open it back up and see if they pick it up. Thanks!
